File: Luhn/ruby/Luhn_solved.rb
def valid?(cc)
  csum = 0
  cc.digits.each_slice(2) do |odd, even|
    double = even.to_i * 2
    double -= 9 if double > 9
    csum += double + odd
  end
  (csum % 10).zero?
end

require 'minitest/autorun'
class TestLuhn < Minitest::Test
  def test_49927398716
    assert valid?(49_927_398_716)
  end

  def test_49927398717
    refute valid?(49_927_398_717)
  end

  def test_1234567812345678
    refute valid?(1_234_567_812_345_678)
  end

  def test_1234567812345670
    assert valid?(1_234_567_812_345_670)
  end
end

File: Luhn/elixir/Luhn_solved.exs
defmodule Luhn do
  def valid?(cc) when is_binary(cc), do: String.to_integer(cc) |> valid?
  def valid?(cc) when is_integer(cc) do
    0 == Integer.digits(cc)
         |> Enum.reverse
         |> Enum.chunk_every(2, 2, [0])
         |> Enum.reduce(0, fn([odd, even], sum) -> Enum.sum([sum, odd | Integer.digits(even*2)]) end)
         |> rem(10)
  end
end

ExUnit.start()
defmodule LuhnTest do
  use ExUnit.Case, async: true

  test "49927398716" do
    assert Luhn.valid?(49927398716)
  end
  test "49927398717" do
    refute Luhn.valid?(49927398717)
  end
  test "1234567812345678" do
    refute Luhn.valid?(1234567812345678)
  end
  test "1234567812345670" do
    assert Luhn.valid?(1234567812345670)
  end
end
